### PR TITLE
[PJRT] Experimental support for `torch.distributed` and DDP on TPU v2/v3

### PR DESCRIPTION
**Pull request by will-cromar**
- Use experimental [`ThreadLocalWorld`](https://github.com/pytorch/pytorch/blob/801d831d7a83c3da41b5d457616cef08769d7281/torch/testing/_internal/distributed/multi_threaded_pg.py#L301-L335) to enable multithreading
- Implement a new `torch.distributed` `init_method` that uses PJRT runtime parameters and supports multithreading
- `torch.distributed` "rank" will become the same as our "ordinal", meaning we have one fewer set of indices to track
- Converge the PJRT ImageNet and MNIST tests with the original ones
- Remove experimental `pjrt.DistributedDataParallel` now that the upstream version works on v3

Performance comparison using ResNet50 with fake data on TPU v3:

Runtime | DDP | Throughput (ex/sec/replica)
-- | -- | --
XRT | No | 418.54
XRT | Yes | 395.97
PJRT | No | ~640
PJRT | Yes | ~565

Example usage:

```python
import torch
import torch.distributed as dist
import torch_xla.core.xla_model as xm
import torch_xla.distributed.xla_multiprocessing as xmp
import torch_xla.experimental.pjrt_backend
from torch_xla.experimental import pjrt

def _all_gather(index: int):
  dist.init_process_group('xla', init_method='pjrt://')
  t = torch.tensor([index], dtype=torch.int32, device=xm.xla_device())
  output = [torch.zeros_like(t) for _ in range(dist.get_world_size())]
  dist.all_gather(output, t)
  xm.mark_step()
  print(output)

if __name__ == '__main__':
  xmp.spawn(self._all_gather)
```

Needs rebasing after #4504 merges

Follow-up:
- Update automated tests
- Update PJRT and DDP documentation before release

### CODE REVIEW

**Comment by JackCaoG** [comment]
@will-cromar is this one ready for review?

**Comment by will-cromar** [comment]
I'll take another pass tomorrow to polish and add some comments, but this should be ready for review.

File: test/pjrt/test_ddp.py [hunk 1]
```diff
@@ -18,39 +17,29 @@
...
+    pjrt.spawn_threads(self._ddp_init)
 
-  @parameterized.named_parameters(*ddp_parameters)
-  def test_ddp_correctness_large_net(self, ddp: type):
+  @parameterized.named_parameters(('small_net', False), ('large_net', True))
+  def test_ddp_correctness(self, use_large_net: bool):
     pjrt._run_multiprocess(
-        util.ddp_correctness, ddp=ddp, use_large_net=True, debug=FLAGS.debug)
+        util.ddp_correctness,
+        init_method='pjrt://',
```
Comment (alanwaketan) [inline test/pjrt/test_ddp.py:40]: Wonder if we want to parameterized the init_method with env as well?

File: test/distributed_util.py [hunk 2]
```diff
@@ -128,10 +129,11 @@ def ddp_correctness(ddp: type = torch.nn.parallel.DistributedDataParallel,
   offset = rank * local_batch_size
   for step in range(steps):
     # To make torch.randn produce same results across devices.
-    torch.manual_seed(2022 + step)
+    rng = torch.Generator().manual_seed(2022 + step)
```
[the remark this answers is not in the document]
Comment (will-cromar) [reply test/distributed_util.py:132]: The other option would be to wrap these `randn` calls in a lock and give them a common global seed, but explicitly creating a new generator with the same seed seems clearer to me. I would have done the same for module initialization, but that case doesn't support a custom RNG.

File: test/pjrt/test_ddp.py (hunk 1 above)
Comment (will-cromar) [reply test/pjrt/test_ddp.py:41]: Good idea. Added another test that skips for TPU <= v3, since `env://` doesn't work nicely with multithreading.

**Review by alanwaketan — APPROVED** [review]
LGTM.

File: test/distributed_util.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alanwaketan) [reply test/distributed_util.py:132]: One last comment. Maybe add a comment to the test case to suggest the reasoning behind?

File: test/test_train_mp_imagenet.py [hunk 3]
```diff
@@ -31,7 +31,7 @@
     '--ddp': {
         'action': 'store_true',
     },
-    '--ddp_pjrt': {
+    '--pjrt_distributed': {
```
Comment (JackCaoG) [inline test/test_train_mp_imagenet.py:34]: can we add comment here to explain how is `pjrt_distributed` different than `ddp`?

**Review by JackCaoG — APPROVED** [review]
Thanks!
